aptos-move/block-executor/src/executor.rs

```diff
@@ -257,6 +257,68 @@ where
         Ok(())
     }
 
+    fn process_delayed_field_output(
+        maybe_output: Option<&E::Output>,
+        idx_to_execute: TxnIndex,
+        read_set: &mut CapturedReads<T, ModuleId, CompiledModule, Module, AptosModuleExtension>,
+        last_input_output: &TxnLastInputOutput<T, E::Output, E::Error>,
+        versioned_cache: &MVHashMap<T::Key, T::Tag, T::Value, DelayedFieldID>,
+    ) -> Result<(), PanicError> {
+        let mut prev_modified_delayed_fields = last_input_output
+            .delayed_field_keys(idx_to_execute)
+            .map_or_else(HashSet::new, |keys| keys.collect());
+
+        // TODO[agg_v2](optimize): see if/how we want to incorporate DeltaHistory from read set into
+        // versioned_delayed_fields. Without it, currently, materialized reads cannot check history
+        // and fail early.
+        //
+        // We can extract histories with something like the code below, and then include history in
+        // change.into_entry_no_additional_history().
+        //
+        // for id in read_set.get_delayed_field_keys() {
+        //     if !delayed_field_change_set.contains_key(id) {
+        //         let read_value = read_set.get_delayed_field_by_kind(id, DelayedFieldReadKind::Bounded).unwrap();
+        //     }
+        // }
+
+        if let Some(output) = maybe_output {
+            for (id, change) in output.delayed_field_change_set().into_iter() {
+                prev_modified_delayed_fields.remove(&id);
+
+                let entry = change.into_entry_no_additional_history();
+
+                // TODO[agg_v2](optimize): figure out if it is useful for change to update needs_suffix_validation
+                if let Err(e) =
+                    versioned_cache
+                        .delayed_fields()
+                        .record_change(id, idx_to_execute, entry)
+                {
+                    match e {
+                        PanicOr::CodeInvariantError(m) => {
+                            return Err(code_invariant_error(format!(
+                                "Record change failed with CodeInvariantError: {:?}",
+                                m
+                            )));
+                        },
+                        PanicOr::Or(_) => {
+                            read_set.capture_delayed_field_read_error(&PanicOr::Or(
+                                MVDelayedFieldsError::DeltaApplicationFailure,
+                            ));
+                        },
+                    };
+                }
+            }
+        }
+
+        for id in prev_modified_delayed_fields {
+            versioned_cache
+                .delayed_fields()
+                .remove_v2(&id, idx_to_execute);
```

Comment:
Currently, this V2 just calls into V1 API, but should we just keep single remove? Otherwise if V2 impl changes here, since V1 still calls into this function we end up doing something different?

Reply:
oh good catch, remove vs remove_v2 is actually just one flag setting, and now V1 was no longer calling remove but remove_v2, meaning it would also have disabled the estimates optimization for V1.. Indeed, maybe I'll just have a single shared remove, and pass the is_blockstm_v2 flag, or always set the flag (since V2 will ignore it anyway)